docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,272 @@
+//      \forall Is:List{Int} . fresh(Is) -> * \exists I:Int . \neg( I \in Is) /\ I
+```
+
+### Frontend Syntax
+
+We use the following K syntax to define `fresh(Is)`
+
+```
+syntax Exp ::= fresh (List{Int})
+rule fresh(Is:List{Int}) => ?I:Int
+  ensures notBool (?I inList{Int} Is)
+```
+
+## Example C: Arbitrary Number (Unspecific Function) `arb()`
+
+The function `arb()` is not a PL construct, but a mathematical function.
+Therefore, its axioms do not use rewrites but equalities.
+
+### Axioms
+
+The intended semantics of `arb()` is that it is an unspecified nullary function.
+The exact return value of `arb()` is unspecified in the semantics but up to the
+implementations.
+However, being a mathematical function, `arb()` must return the same value in
+any one implementation.
+
+The axiom of `arb()` is
+
+```
+(ARB) \exists I:Int . arb() = I
+//    this is exactly the (Function) axiom
+//    this is not equivalent to
+//      arb() = (\exists I:Int . I)
+```
+
+There are many variants of `arb()`. For example, `arbDifferentFromList(Is)` is
+a unspecified function whose return value must be different than those in `Is`.
+Here is its axiom
+
+```
+(ARBDIFF) \forall Is:List{Is} . \exists I:Int . \neg(I \in Is) /\ arbDifferentFromList(Is) = I
+```
+
+Note that `arbDifferentFromList(Is)` is different from `fresh(Is)`, because
+`fresh(Is)` _transitions_ to an integer not in `Is` (could be a different one
+each time it is used), while `arbDifferentFromList(Is)` _is equal to_ a (fixed)
+integer not in `Is`.
+
+### Skolemization
+
+One possible approach to implementing the abovewould be through [Skolemization](https://en.wikipedia.org/wiki/Skolem_normal_form),
+i.e., by replacing the existential variables with fresh uninterpreted functions
+applied to the universally quantified variables.
+
+For example, the `(ARBDIFF)` rule above could be Skolemized to
+
+```
+(ARBDIFF') \forall Is:List{Is} . \neg(arbDiffI(Is) \in Is) /\ arbDifferentFromList(Is) = arbDiffI(Is)
+```
+where
+```
+arbDiffI : List{Int} -> Int
+```
+
+is a symbol satisfying the `(Function)` axiom, without further constraints.
+
+### Frontend Syntax
+
+We do not need special frontend syntax to define `arb()`.
+We only need to define it in the usual way as a function
+(instead of a language construct), and provide no axioms for it
+(the `(Function)` axiom is generated automatically from the `function`
+attribute annotation.
+
+The frontend doesn't currently provide a syntax to define the variants of
+`arb()`, such as `arbDifferentFromList(Is)` in the above.
+
+We envision two possible solutions:
+
+1. they are not allowed, since the users could themselves Skolemize the axiom
+
+    ```
+    syntax Int ::= arbDiffI(List{Int}) [function, functional]
+    syntax Int ::= arbDifferentFromList(List{Int}) [function, functional]
+    rule arbDifferentFromList(Is:List{Int}) => arbDiffI(Is)
+      ensures notBool (arbDiffI(Is) inList{Int} Is)
+    ```
+
```

Comment:
Allowing both `function` and `functional` as attributes is really confusing.  We should change one of them.

Reply:
Also, the rule above looks strange to me.  `arbDiffI(Is)` seems artificial there.

Reply:
I removed this and only kept the `?` syntax and promised an Skolemization pass in the Kompiler